shell/platform/darwin/ios/framework/Headers/FlutterPluginAppLifeCycleDelegate.h

```diff
@@ -66,7 +66,7 @@ FLUTTER_DARWIN_EXPORT
 /**
  * Calls all plugins registered for `UIApplicationDelegate` callbacks.
  */
-- (void)application:(UIApplication*)application
+- (void)performApplication:(UIApplication*)application
```

Comment:
Wouldn't this be a breaking change for plugins that have implemented this method?

What about `application:didRegisterForRemoteNotificationsWithDeviceToken`? `didFailToRegisterForRemoteNotificationsWithError`? etc
https://github.com/firebase/firebase-ios-sdk/blob/7eb1747e4b9be195ea9908b9da28b1d0f818afe3/GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m#L111-L112

Reply:
> Wouldn't this be a breaking change for plugins that have implemented this method?

I'm not sure what you have in mind.  This is on the FlutterPluginAppLifeCycleDelegate.  Do you mean the usage of `_cmd`?  The modification of the class at runtime takes care of that as demonstrated in the test.

> application:didRegisterForRemoteNotificationsWithDeviceToken

Yea, I'll do that one too, thanks.

Reply:
> Wouldn't this be a breaking change for plugins that have implemented this method?

If a plugin has `application:didReceiveRemoteNotification:fetchCompletionHandler:` implemented like this:
https://github.com/blackmenthor/flutter-branch-io/blob/39746a23dcacca266cc8af98efda7687643bed89/ios/Classes/SwiftFlutterBranchIoPlugin.swift#L173-L176

It's not implementing `performApplication:didReceiveRemoteNotification:fetchCompletionHandler:`, but the implementation has been swapped out in `+load`.  Would their implementation be called?  Is there a way we can make sure that delegates that already respond to `application:didReceiveRemoteNotification:fetchCompletionHandler:` don't break?

Reply:
Hmm, let me write up some tests for that use case. It's hard to know what will happen without trying it.

I've added `application:didRegisterForRemoteNotificationsWithDeviceToken` and backfilled tests to FlutterAppDelegate.

Reply:
Firebase also did `didFailToRegisterForRemoteNotificationsWithError`  I'll do that one too.

Reply:
Okay, so I looked into the use case for subclassing `FlutterPluginAppLifeCycleDelegate` like in the example provided. 
 It is using the `override` annotation so removing `application:didReceiveRemoteNotification:fetchCompletionHandler:` will be a problem since its removal is necessary in order to fix the problem.  I've added a test to show that subclassing and overriding `performApplication:didReceiveRemoteNotification:fetchCompletionHandler:` behaves as expected.  So this plugin will break in the best way possible, at compilation time when the `override` annotation doesn't make any sense.  They'll have to rename their method to have the `perform` prefix.  For objc subclasses this shouldn't be a problem since there is no "override" analog.   I don't believe this is technically a breaking change since it doesn't break any tests.  I'm not sure if this is different criteria for plugins.

Reply:
Cross-referencing https://github.com/firebase/firebase-ios-sdk/pull/2835

> So this plugin will break in the best way possible, at compilation time when the override annotation doesn't make any sense.

Fortunately I doubt this feature is used by many plugins.

> For objc subclasses this shouldn't be a problem since there is no "override" analog.

The problem there is that it wouldn't break, and they wouldn't know they need to rename the method.  Could we detect that case in `+load` and `NSAssert`?  Or just avoid the swapping?

> I don't believe this is technically a breaking change since it doesn't break any tests

A remote notification test would be difficult, but if we had one it would break, but point taken.

Reply:
Added didFailToRegisterForRemoteNotificationsWithError.

Reply:
> The problem there is that it wouldn't break, and they wouldn't know they need to rename the method. Could we detect that case in +load and NSAssert? Or just avoid the swapping?

For objc the hypothetical subclasser wouldn't have to rename it, it would just work if the implementation is like the example provided above.  If they are calling `[super application:didReceiveRemoteNotification:fetchCompletionHandler:]` in the subclass it will create a warning or error depending on their compilation settings but would execute fine.